java/src/main/java/ai/onnxruntime/OnnxTensor.java

```diff
@@ -780,32 +682,91 @@ static OnnxTensor createTensor(
-      return new OnnxTensor(
-          createTensorFromBuffer(
-              OnnxRuntime.ortApiHandle,
-              allocator.handle,
-              tmp,
-              bufferSize,
-              shape,
-              info.onnxType.value),
-          allocator.handle,
-          info,
-          tmp);
+      return createTensor(type, allocator, data, shape);
     } else {
       throw new IllegalStateException("Trying to create an OnnxTensor on a closed OrtAllocator.");
     }
   }
 
+  /**
+   * Creates a tensor by wrapping the data in a direct byte buffer and passing it to JNI.
+   *
+   * <p>Throws IllegalStateException if the buffer is too large to create a direct byte buffer copy,
+   * which is more than approximately (Integer.MAX_VALUE - 5) / type.size elements.
+   *
+   * @param type The buffer type.
+   * @param allocator The OrtAllocator.
+   * @param data The data.
+   * @param shape The tensor shape.
+   * @return An OnnxTensor instance.
+   * @throws OrtException If the create call failed.
+   */
+  private static OnnxTensor createTensor(
+      OnnxJavaType type, OrtAllocator allocator, Buffer data, long[] shape) throws OrtException {
+    int bufferPos;
+    int bufferSize = data.remaining() * type.size;
+    if ((bufferSize < 0) || (bufferSize > ((Integer.MAX_VALUE / type.size) - 10))) {
+      // overflowed as we can't make a direct byte buffer of that size
+      throw new IllegalStateException(
+          "Cannot allocate a direct buffer of the requested size and type, size "
+              + data.remaining()
+              + ", type = "
+              + type);
+    }
+    Buffer tmp;
+    if (data.isDirect()) {
+      tmp = data;
+      bufferPos = data.position() * type.size;
+    } else {
+      // Copy the data to a new direct buffer, then restore the state of the input.
+      int origPosition = data.position();
+      ByteBuffer buffer = ByteBuffer.allocateDirect(bufferSize).order(ByteOrder.nativeOrder());
+      switch (type) {
+        case FLOAT:
+          tmp = buffer.asFloatBuffer().put((FloatBuffer) data);
+          break;
+        case DOUBLE:
+          tmp = buffer.asDoubleBuffer().put((DoubleBuffer) data);
+          break;
+        case INT8:
+          // buffer is already a ByteBuffer, no cast needed.
+          tmp = buffer.put((ByteBuffer) data);
+          break;
+        case INT16:
+          tmp = buffer.asShortBuffer().put((ShortBuffer) data);
+          break;
+        case INT32:
+          tmp = buffer.asIntBuffer().put((IntBuffer) data);
+          break;
+        case INT64:
+          tmp = buffer.asLongBuffer().put((LongBuffer) data);
+          break;
+        case BOOL:
+        case STRING:
+        case UNKNOWN:
+        default:
+          throw new IllegalStateException(
+              "Impossible to reach here, managed to cast a buffer as an incorrect type");
+      }
+      data.position(origPosition);
+      tmp.rewind();
+      bufferPos = 0;
```

Comment:
>bufferPos = 0; [](start = 6, length = 14)

Not clear why bytes position is not what it is for direct buffer?

Reply:
Because this is a fresh buffer, with the elements copied from the starting position of the input buffer `data`, but sized so it's only got `data.remaining()` capacity. Basically it's copying it into a fresh array, so our index starts from 0 instead of `data.position()`.